### PR TITLE
Use cub::DeviceScan to implement thrust scans on CUDA.

### PR DESCRIPTION
**Pull request by allisonvacanti**
Fixes: #1301 

Prerequisite: NVIDIA/cub#210

### CODE REVIEW

**Comment by allisonvacanti** [comment]
See also https://github.com/NVIDIA/thrust/pull/1304

**Comment by allisonvacanti** [comment]
@griwes Can you check that this is properly using the index type dispatch macros? I want to double check that I'm not missing anything.

**Comment by allisonvacanti** [comment]
DVS CL 29194365

**Comment by allisonvacanti** [comment]
This is causing DVS to time out. Will try again after NVIDIA/cub#213 is merged.

File: thrust/system/cuda/detail/scan.h [hunk 1]
```diff
@@ -26,762 +26,204 @@
  ******************************************************************************/
 #pragma once
 
-
-#if THRUST_DEVICE_COMPILER == THRUST_DEVICE_COMPILER_NVCC
```
Comment (brycelelbach) [inline thrust/system/cuda/detail/scan.h:30]: Let's leave this in for now. @jaredhoberock @griwes does anyone know the historical reason why we had to wrap the CUDA backend headers in `#if THRUST_DEVICE_COMPILER == THRUST_DEVICE_COMPILER_NVCC`?

**Review by brycelelbach — APPROVED** [review]

File: thrust/system/cuda/detail/scan.h [hunk 2]
```diff
@@ -26,762 +26,204 @@
...
+                                 THRUST_DEBUG_SYNC_FLAG));
+    thrust::cuda_cub::throw_on_error(status,
+                                     "after determining tmp storage "
+                                     "requirements for exclusive_scan");
+  }
 
-    THRUST_INDEX_TYPE_DISPATCH(status,
-                                doit_step<Inclusive>,
+  // Run scan:
+  {
```
Comment (brycelelbach) [inline thrust/system/cuda/detail/scan.h:173]: This style is fine, but we should do it in reduce and sort as well.

File: thrust/system/cuda/detail/scan.h (hunk 1 above)
Comment (allisonvacanti) [reply thrust/system/cuda/detail/scan.h:30]: Done.

**Comment by allisonvacanti** [comment]
Can't get DVS to finish, we're hitting a 25(!) minute build timeout for CL 29265544.

**Comment by allisonvacanti** [comment]
Rebased on master to resolve submodule conflict.

DVS CL: 29371523

**Comment by allisonvacanti** [comment]
DVS is using an old version of MSVC 2019 (19.20) that is timing out when building `transform_scan.cu`.

@brycelelbach can you update that config to use a newer version of the compiler? Looks like VS 16.8 (`cl` 19.28) is available on DVS, that's what I use locally without issues.

**Comment by allisonvacanti** [comment]
Checking whether the timeouts are still an issue with CL 29471041.

**Pull request merged**